src/main/java/hudson/tasks/junit/History.java

```diff
@@ -140,42 +552,58 @@ public boolean isDescriptionAvailable() {
-        int offset = userOffset;
-        if (userOffset > 1000 || userOffset < 0) {
-            offset = 0;
+        public String getTrendChartJson() {
+            return trendChartJson;
         }
 
-        TestResultImpl pluggableStorage = getPluggableStorage();
+        public String getTrendChartJsonStr() {
+            return trendChartJsonStr;
+        }
+    }
 
+    public HistoryTableResult retrieveHistorySummary(int start, int end) {
+        TestResultImpl pluggableStorage = getPluggableStorage();
+        List<HistoryTestResultSummary> trs = null;
         if (pluggableStorage != null) {
-            return new HistoryTableResult(pluggableStorage.getHistorySummary(offset));
+            int offset = start;
+            if (start > 1000 || start < 0) {
+                offset = 0;
+            }
+            trs = pluggableStorage.getHistorySummary(offset);
+        } else {
+            trs = getHistoryFromFileStorage(start, end);
         }
-        return new HistoryTableResult(getHistoryFromFileStorage());
+        return new HistoryTableResult(trs, computeTrendJsons(trs));
     }
-
-    private List<HistoryTestResultSummary> getHistoryFromFileStorage() {
+    static int parallelism = Math.min(Runtime.getRuntime().availableProcessors(), Math.max(4, (int)(Runtime.getRuntime().availableProcessors() * 0.75 * 0.75)));
+    static ExecutorService executor = Executors.newFixedThreadPool(Math.max(4, (int)(Runtime.getRuntime().availableProcessors() * 0.75 * 0.75)));
+    private List<HistoryTestResultSummary> getHistoryFromFileStorage(int start, int end) {
         TestObject testObject = getTestObject();
         RunList<?> builds = testObject.getRun().getParent().getBuilds();
-        return builds
-                .stream()
-                .map(build -> {
-                    hudson.tasks.test.TestResult resultInRun = testObject.getResultInRun(build);
-                    if (resultInRun == null) {
-                        return null;
-                    }
-
-                    return new HistoryTestResultSummary(build, resultInRun.getDuration(),
-                            resultInRun.getFailCount(),
-                            resultInRun.getSkipCount(),
-                            resultInRun.getPassCount(),
-                            resultInRun.getDescription()
-                    );
-                })
-                .filter(Objects::nonNull)
-                .collect(Collectors.toList());
+        final AtomicInteger count = new AtomicInteger(0);
+        final long startedMs = java.lang.System.currentTimeMillis();
+        return builds.stream().skip(start)
+            .collect(ParallelCollectors.parallel(build -> {
+                int c = count.incrementAndGet();
+                if (c > end - start + 1 || (java.lang.System.currentTimeMillis() - startedMs) > 15000) { // Do not navigate too far or for too long, we need to finish the request this year and have to think about RAM
```

Comment:
extract the 15000 out to a constant, something like MAX_TIME_ELAPSED_RETRIEVING_RESULTS